engine/src/main/java/org/terasology/rendering/dag/RenderTaskListGenerator.java

```diff
@@ -56,47 +54,57 @@ private void generateTaskList() {
+                            taskList.add(stateChange.generateTask());
+                            persistentStateChanges.put(stateChange.getClass(), stateChange);
+
+                        } // else {
+                        // do nothing: we do not want a back-to-default state change if there was no known
+                        // non-default state change already taking place.
+                        // }
+                    } else {
+                        if (stateChange.isTheDefaultInstance()) { // I know, I'm a sucker for almost plain-english code
+                            // non redundant default state change, eliminates subType entry in the map, to keep map small
+                            taskList.add(stateChange.generateTask());
+                            persistentStateChanges.remove(stateChange.getClass());
+
+                        } else if (!stateChange.isEqualTo(persistentStateChange)) { // another new method, just for readability
+                            // non-redundant state change of the same subType but different value, becomes new map entry
+                            taskList.add(stateChange.generateTask());
+                            persistentStateChanges.put(stateChange.getClass(), stateChange);
+
+                        } // else {
+                        // the non-redundant state change being examined is not a default one and is identical
+                        // in value to the one stored in the persistentStateChange map: it is redundant after all
+                        // and we ignore it
+                        // }
+                    }
 
-                    semiReducedStateChanges.add(entry.getValue());
-                    iterator.remove();
                 }
 
-                semiReducedStateChanges.add(object);
-            }
-        }
-
-        // Add remaining state changes
-        for (Object stateChange : intranodesStateChanges.values()) {
-            semiReducedStateChanges.add(stateChange);
-        }
-
-        Map persistentStateChanges = Maps.newHashMap();
-        for (Object object : semiReducedStateChanges) {
-            if (object instanceof StateChange) {
-                StateChange stateChange = (StateChange) object;
-                if (!persistentStateChanges.containsKey(object.getClass())) {
-                    persistentStateChanges.put(object.getClass(), object);
-                    taskList.add(stateChange.generateTask());
-                } else {
-                    StateChange persistenStateChange = (StateChange) persistentStateChanges.get(object.getClass());
-                    if (!persistenStateChange.compare(stateChange)) {
-                        persistentStateChanges.put(stateChange.getClass(), stateChange);
-                        taskList.add(stateChange.generateTask());
-                    }
-                }
-            } else {
+                intranodesStateChanges.clear();
```

Comment:
Yes, just leaves resetting to default step, which is not desired as discussed above. I tuned to tests and they're passing.